### PR TITLE
Add matching on sort qualities in RR and fix #18787

### LINKED ISSUE #18787: Issues with rewrite rules
### Description of the problem

Three issues with rewrite rules:
- Anomaly `Invalid_argument("Array.fold_left3_i")` on cbv
- `Syntax error: '==>' expected after [rw_pattern] (in [rewrite_rule]).` when ssrbool is imported
- `Unknown existential variable.` if pattern variable is in a case branch

### Small Coq file to reproduce the bug

```coq
(* Issue 1 *)
Symbol dom : Prop -> Prop.
Rewrite Rule dom_rew := dom (forall (x : ?A), ?B) ==> ?A.

(* Issue 2 *)
From Coq Require Import ssrbool.
Symbol a : unit.
Rewrite Rule a_rew := a ==> tt.

(* Issue 3 *)
Symbol t : unit.
Rewrite Rule t_rew :=
| match t with tt => @eq_refl ?A ?a end ==> @eq_refl ?A ?a.
```

### Version of Coq where this bug occurs

8.20+alpha

### Last version of Coq where the bug did not occur

_No response_

### PR DESCRIPTION
**Pull request by yannl35133**
Adds matching on sort qualities in rewrite rules and
Fixes / closes #18787
This changes the rewrite rule syntax : rules look like `udecl |- pat >-> replacement`

- [x] Added / updated **documentation**.
  - [x] Updated **documented syntax** by running `make doc_gram_rsts`.

### CODE REVIEW

File: vernac/g_vernac.mlg [hunk 1]
```diff
@@ -502,7 +502,7 @@ GRAMMAR EXTEND Gram
     [ [ p = lconstr -> { p } ] ];
   rewrite_rule:
     [ [ u = OPT [ u = univ_decl; "|-" -> { u }];
-        lhs = rw_pattern; "==>" ; rhs = lconstr ->
+        lhs = rw_pattern; ">->" ; rhs = lconstr ->
```
Comment (jfehrle) [inline vernac/g_vernac.mlg:505]: Won't this break existing user code that relies on this?  Please address that--either explain why that isn't the case or describe the deprecation/compatibility strategy in the PR.

Comment (jfehrle) [reply vernac/g_vernac.mlg:505]: Ah, this was just added, so not an issue?

Comment (SkySkimmer) [reply vernac/g_vernac.mlg:505]: >Ah, this was just added, so not an issue?

yes

**Comment by SkySkimmer** [comment]
PR was split

**Pull request closed without merging**